### PR TITLE
feat: add sequence-based auto-updates non-semver apps

### PR DESCRIPTION
**Pull request by DanStough**
Signed-off-by: Kira Boyle <kira@replicated.com>

#### What type of PR is this?
type::feature

#### What this PR does / why we need it:
The behavior of automatic-deployments for non-semver channels still followed the same sorting logic as those with semantic versioning enabled. This feature uses sequence-based updates for those apps. It also cleans up the modal dialog to make user selections clearer.

Screenshots of Automatica Updates Modal:

Semver Enabled:
image

Semver Disabled:
image

#### Which issue(s) this PR fixes:
Fixes [SC-44111](https://app.shortcut.com/replicated/story/44111/automatic-upgrades-should-sequence-based-on-semvers-if-the-channel-is-semver-enabled-otherwise-they-should-sort-based-on)

#### Does this PR introduce a user-facing change?
```release-note
- Application release channels that do not have [semantic versioning](https://docs.replicated.com/vendor/releases-understanding#semantic-versioning) enabled can now perform automatic release deployments. The most recent release will be used when updating, regardless of its version tag.
```

#### Does this PR require documentation?
Yes, TBD.

### CODE REVIEW

File: pkg/updatechecker/updatechecker.go [hunk 1]
```diff
@@ -449,8 +449,9 @@ func deployVersionLabel(opts CheckForUpdatesOpts, clusterID string, versionLabel
 	return nil
 }
 
-func autoDeploy(opts CheckForUpdatesOpts, clusterID string, semverAutoDeploy apptypes.SemverAutoDeploy) error {
-	if semverAutoDeploy == "" || semverAutoDeploy == apptypes.SemverAutoDeployDisabled {
+func autoDeploy(opts CheckForUpdatesOpts, clusterID string, autoDeploy apptypes.AutoDeploy) error {
+	if autoDeploy == "" || autoDeploy == apptypes.AutoDeployDisabled {
+		fmt.Println("auto deploy not running")
```
Comment (divolgin) [inline pkg/updatechecker/updatechecker.go:454]: debug stuff?

**Review by divolgin — DISMISSED** [review]

File: pkg/updatechecker/updatechecker.go (hunk 1 above)
Comment (DanStough) [reply pkg/updatechecker/updatechecker.go:454]: yerp!

File: pkg/updatechecker/updatechecker.go [hunk 2]
```diff
@@ -469,33 +469,40 @@ func autoDeploy(opts CheckForUpdatesOpts, clusterID string, semverAutoDeploy app
...
 	var versionToDeploy *downstreamtypes.DownstreamVersion
 
-Loop:
-	for _, v := range appVersions.AllVersions {
-		if v == nil || v.Semver == nil {
-			continue
-		}
+	// In the case semver is not required, we only need to check if the newest app version
+	// is greater than the current version.
+	if autoDeploy == apptypes.AutoDeploySequence && currentVersion.Sequence < appVersions.AllVersions[0].Sequence {
```
Comment (sgalsaleh) [inline pkg/updatechecker/updatechecker.go:474]: what if `autoDeploy` is set to `apptypes.AutoDeploySequence` but the current sequence is greater than or equal to the latest version sequence? then it's gonna execute the semantic versioning logic below instead, which i think is incorrect.

File: web/src/components/apps/AppVersionHistory.jsx [hunk 3]
```diff
@@ -1455,8 +1488,9 @@ class AppVersionHistory extends Component {
               isOpen={this.state.showAutomaticUpdatesModal}
               onRequestClose={this.toggleAutomaticUpdatesModal}
               updateCheckerSpec={app?.updateCheckerSpec}
-              semverAutoDeploy={app?.semverAutoDeploy}
+              autoDeploy={app?.autoDeploy}
               appSlug={app?.slug}
+              semVerEnabled={appLicense?.isSemverRequired}
```
Comment (sgalsaleh) [inline web/src/components/apps/AppVersionHistory.jsx:1493]: this should be part of the application response object, we shouldn't be making an extra request to get the license and determine that.

Comment (sgalsaleh) [reply web/src/components/apps/AppVersionHistory.jsx:1493]: in here i mean https://github.com/replicatedhq/kots/blob/a606fcfa00ba79894376f43ab96053c3b11a7c43/pkg/handlers/app.go#L165

File: web/src/components/modals/AutomaticUpdatesModal.jsx [hunk 4]
```diff
@@ -142,15 +164,28 @@ export default class AutomaticUpdatesModal extends React.Component {
...
-  handleSemverAutoDeployOptionChange = selectedSemverAutoDeploy => {
+  handleAutoDeployOptionChange = selectedAutoDeploy => {
     this.setState({
-      selectedSemverAutoDeploy: { ...selectedSemverAutoDeploy },
+      selectedAutoDeploy: { ...selectedAutoDeploy },
     });
   }
 
+  handleSequenceAutoUpdatesChange = (sequenceAutoDeployEnabled) => {
+    console.log("sequenceAutoDeployEnabled", sequenceAutoDeployEnabled);
```
Comment (sgalsaleh) [inline web/src/components/modals/AutomaticUpdatesModal.jsx:174]: remove log

File: web/src/components/modals/AutomaticUpdatesModal.jsx [hunk 5]
```diff
@@ -214,18 +249,42 @@ export default class AutomaticUpdatesModal extends React.Component {
...
-              classNamePrefix="replicated-select"
-              placeholder="Automatically deploy new versions"
-              options={SEMVER_AUTO_DEPLOY_OPTIONS}
-              isSearchable={false}
-              getOptionValue={(option) => option.label}
-              value={selectedSemverAutoDeploy}
-              onChange={this.handleSemverAutoDeployOptionChange}
-              isOptionSelected={(option) => { option.value === selectedSemverAutoDeploy }}
-              />
+            { semVerEnabled ?
```
Comment (sgalsaleh) [inline web/src/components/modals/AutomaticUpdatesModal.jsx:252]: how about having the variable names be consistent? `semVerEnabled` vs `isSemverRequired`

Comment (sgalsaleh) [reply web/src/components/modals/AutomaticUpdatesModal.jsx:252]: i would've thought they're two different things, but it seems like they're the same?

File: web/src/components/apps/AppVersionHistory.jsx (hunk 3 above)
Comment (DanStough) [reply web/src/components/apps/AppVersionHistory.jsx:1493]: I'll have to add it to the App Response, but I see there are other license fields in their too. I'll consolidate that.

File: pkg/updatechecker/updatechecker.go (hunk 2 above)
Comment (DanStough) [reply pkg/updatechecker/updatechecker.go:474]: Yeah, great callout.

File: web/src/components/modals/AutomaticUpdatesModal.jsx (hunk 5 above)
Comment (DanStough) [reply web/src/components/modals/AutomaticUpdatesModal.jsx:252]: Yes, they are the same. I will change to `isSemverRequired`

**Review by sgalsaleh — DISMISSED** [review]

**Review by sgalsaleh — APPROVED** [review]

**Pull request merged**